cli/src/doctest.rs

```diff
@@ -0,0 +1,417 @@
+        let term = program.parse()?;
+        let cache = &mut program.vm.import_resolver;
+        let transformed = doctest_transform(cache, &mut registry, term);
+        cache.set(program.main_id, transformed.clone()?, EntryState::Parsed);
+        Ok((program.eval_record_spine()?, registry))
+    }
+}
+
+fn nickel_code_blocks<'a>(document: &'a AstNode<'a>) -> Vec<DocTest> {
+    use comrak::arena_tree::Node;
+    use comrak::nodes::{Ast, NodeCodeBlock, NodeValue};
+    document
+        .traverse()
+        .flat_map(|ne| match ne {
+            // Question: can we extract enough location information so that
+            // we can munge the parsed AST to point into the doc comment?
+            NodeEdge::Start(Node { data, .. }) => match &*data.borrow() {
+                Ast {
+                    value: NodeValue::CodeBlock(NodeCodeBlock { info, literal, .. }),
+                    ..
+                } => info
+                    .strip_prefix("nickel")
+                    .map(|tag| match tag.trim() {
+                        "ignore" => Vec::new(),
+                        "multiline" => {
+                            static BLANK_LINE: Lazy<Regex> =
+                                Lazy::new(|| Regex::new("\n\\s*\n").unwrap());
+                            BLANK_LINE
+                                .split(literal)
+                                .filter_map(|chunk| {
+                                    if !chunk.trim().is_empty() {
+                                        Some(DocTest::new(chunk.to_owned()))
+                                    } else {
+                                        None
+                                    }
+                                })
+                                .collect()
+                        }
+                        _ => vec![DocTest::new(literal.to_owned())],
+                    })
+                    .unwrap_or_default(),
+                _ => vec![],
+            },
+            _ => vec![],
+        })
+        .collect()
+}
+
+// Transform a term by taking all its doctests and inserting them into the record next
+// to the field that they're annotating.
+//
+// For example,
+// {
+//   field | doc m%"
+//     ```nickel
+//       1 + 1
+//     ```
+//   "%
+// }
+// becomes
+// {
+//   field | doc m%"
+//     ```nickel
+//       1 + 1
+//     ```
+//   "%,
+//   %0 = fun %1 => 1 + 1,
+// }
+//
+// The idea is for the test to be evaluated in the same environment as the
+// field that declares it. We wrap the test in a function so that it doesn't get
+// evaluated too soon.
+//
+// One disadvantage with this traversal approach is that any parse errors in
+// the test will be encountered as soon as we explore the record spine. We might
+// prefer to delay parsing the tests until it's time to evaluate them.
+// The main advantage of this approach is that it makes it easy to have the test
+// evaluated in the right environment.
```

Comment:
This documentation doesn't explain its action on the registry.

Reply:
Good point, added.